dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVTfo.mo

```diff
@@ -0,0 +1,88 @@
+
+  input Types.VoltageModulePu UStatorRefPu(start = UStatorRef0Pu) "Voltage regulation set point at stator in pu (base UNom)";
+
+  Types.VoltageModulePu UStatorPu(start = UStator0Pu) "Voltage module at stator in pu (base UNom)";
+  Types.ComplexVoltagePu uStatorPu(re(start = uStator0Pu.re), im(start = uStator0Pu.im)) "Complex voltage at stator in pu (base UNom)";
+  Types.ComplexCurrentPu iStatorPu(re(start = iStator0Pu.re), im(start = iStator0Pu.im)) "Complex current at stator in pu (base UNom, SNom) (generator convention)";
+  Types.ComplexApparentPowerPu sStatorPu(re(start = sStator0Pu.re), im(start = sStator0Pu.im)) "Complex apparent power at stator in pu (base UNom, SNom) (generator convention)";
+  Types.ReactivePowerPu QStatorPu(start = QStator0Pu) "Stator reactive power in pu (base QNomAlt) (generator convention)";
+  Boolean limUQUp(start = false) "Whether the maximum reactive power limits are reached or not (from generator voltage regulator)";
+  Boolean limUQDown(start = false) "Whether the minimum reactive power limits are reached or not (from generator voltage regulator)";
+
+  parameter Types.VoltageModulePu UStatorRef0Pu "Start value of voltage regulation set point at stator in pu (base UNom)";
+  parameter Types.VoltageModulePu UStator0Pu "Start value of voltage module at stator in pu (base UNom)";
+  parameter Types.ComplexVoltagePu uStator0Pu "Start value of complex voltage at stator in pu (base UNom)";
+  parameter Types.ComplexCurrentPu iStator0Pu "Start value of complex current at stator in pu (base UNom, SNom) (generator convention)";
+  parameter Types.ComplexApparentPowerPu sStator0Pu "Start value of complex apparent power at stator in pu (base UNom, SNom) (generator convention)";
+  parameter Types.ReactivePowerPu QStator0Pu "Start value of stator reactive power in pu (base QNomAlt) (generator convention)";
+
+protected
+  QStatus qStatus(start = QStatus.Standard) "Voltage regulation status: standard, absorptionMax or generationMax";
+
+equation
+  when QGenPu <= QMinPu and UStatorPu >= UStatorRefPu then
+    qStatus = QStatus.AbsorptionMax;
+    limUQUp = false;
+    limUQDown = true;
+  elsewhen QGenPu >= QMaxPu and UStatorPu <= UStatorRefPu then
+    qStatus = QStatus.GenerationMax;
+    limUQUp = true;
+    limUQDown = false;
+  elsewhen (QGenPu > QMinPu or UStatorPu < UStatorRefPu) and (QGenPu < QMaxPu or UStatorPu > UStatorRefPu) then
+    qStatus = QStatus.Standard;
+    limUQUp = false;
+    limUQDown = false;
+  end when;
+
+  if running.value then
+    if qStatus == QStatus.GenerationMax then
+      QGenPu = QMaxPu;
+    elseif qStatus == QStatus.AbsorptionMax then
+      QGenPu = QMinPu;
+    else
+      UStatorPu = UStatorRefPu;
+    end if;
+  else
+    terminal.i.im = 0;
+  end if;
+
+  UStatorPu = ComplexMath.'abs'(1 / rTfoPu * (terminal.V - terminal.i * Complex(RTfoPu, XTfoPu) * SystemBase.SnRef / SNom));
```

Comment:
Simpler and more performant to just write UStatorPu = ComplexMath.'abs'(uStatorPu)

Reply:
Ok